File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/three-gpu-computation-handle-init-error.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { findTransparentExpressionRoot } from "../../utils/find-transparent-expression-root.js";
import { getStaticPropertyName } from "../../utils/get-static-property-name.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { getThreeConstructorName } from "./utils/get-three-constructor-name.js";

const isDiscardedExpression = (node: EsTreeNode): boolean => {
  const expressionRoot = findTransparentExpressionRoot(node);
  if (
    expressionRoot.parent &&
    isNodeOfType(expressionRoot.parent, "ExpressionStatement") &&
    expressionRoot.parent.expression === expressionRoot
  ) {
    return true;
  }
  return Boolean(
    expressionRoot.parent &&
    isNodeOfType(expressionRoot.parent, "UnaryExpression") &&
    expressionRoot.parent.operator === "void",
  );
};

export const threeGpuComputationHandleInitError = defineRule({
  id: "three-gpu-computation-handle-init-error",
  title: "GPU computation initialization error is ignored",
  category: "Correctness",
  severity: "error",
  recommendation: "Check or propagate the error string returned by GPUComputationRenderer.init()",
  create: (context: RuleContext) => ({
    CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
      if (
        !isNodeOfType(node.callee, "MemberExpression") ||
        getStaticPropertyName(node.callee) !== "init" ||
        getThreeConstructorName(node.callee.object, context.scopes) !== "GPUComputationRenderer" ||
        !isDiscardedExpression(node)
      ) {
        return;
      }
      context.report({
        node,
        message:
          "GPUComputationRenderer.init() returns null on success or an error string on failure, but this result is discarded",
      });
    },
  }),
});

File: native/oxlint/rules/three-gpu-computation-handle-init-error.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;
use oxc_syntax::operator::UnaryOperator;

use crate::{context::LintContext, rule::Rule, AstNode};

#[derive(Debug, Default, Clone)]
pub struct ThreeGpuComputationHandleInitError;

declare_oxc_lint!(
    /// Require callers to handle GPUComputationRenderer initialization errors.
    ThreeGpuComputationHandleInitError,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Require GPU computation initialization error handling.",
);

impl Rule for ThreeGpuComputationHandleInitError {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::CallExpression(call_expression) = node.kind() else {
            return;
        };
        let Some(member_expression) = call_expression.callee.as_member_expression() else {
            return;
        };
        if member_expression.static_property_name() != Some("init")
            || three_constructor_api_name(member_expression.object(), ctx).as_deref()
                != Some("GPUComputationRenderer")
        {
            return;
        }
        let expression_root = transparent_expression_root(node, ctx);
        let parent = ctx.nodes().parent_node(expression_root.id());
        let is_discarded = matches!(parent.kind(), AstKind::ExpressionStatement(_))
            || matches!(
                parent.kind(),
                AstKind::UnaryExpression(unary_expression)
                    if unary_expression.operator == UnaryOperator::Void
            );
        if !is_discarded {
            return;
        }
        ctx.diagnostic(
            OxcDiagnostic::warn(
                "GPUComputationRenderer.init() returns null on success or an error string on failure, but this result is discarded",
            )
            .with_label(node.span()),
        );
    }
}
